WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift

```diff
@@ -175,6 +199,139 @@ extension PointOfSaleAggregateModel {
             await cardPresentPaymentService.disconnectReader()
         }
     }
+
+    /// Starts a payment immediately if a reader is connected.
+    /// Otherwise, schedules a payment to start the next time a reader connects.
+    /// Note that any schedlued payments are cancelled by `cancelReaderPreparation`
+    /// e.g. when the TotalsView goes offscreen.
+    func startPaymentWhenCardReaderConnected() async {
+        guard case .connected = cardReaderConnectionStatus else {
+            return startPaymentOnCardReaderConnection = $cardReaderConnectionStatus
+                .filter { status in
+                    switch status {
+                    case .connected:
+                        return true
+                    case .disconnected, .disconnecting, .cancellingConnection:
+                        return false
+                    }
+                }
+                .removeDuplicates()
+                .sink { _ in
+                    Task { @MainActor [weak self] in
+                        await self?.collectPayment()
+                    }
+                }
+        }
+        await collectPayment()
+    }
+
+    @MainActor
+    func collectPayment() async {
+        guard let order else {
+            return
+            // Should this throw?
```

Comment:
I think logging is more than enough for now.

There's always a possibility code-wise to enter such states. I'm thinking if in the future one more way to organize the aggregate model into a hierarchy of the lifetime of the states.

e.g.

POS State - life cycle permanent: merchant settings, products, inventory, reader connection, order history, current order...
Order state - life cycle single order: selected items, progress, current payment
Payment state - life cycle single payment: payment progress

The order state expects the POS model as input, and the payment state expects the order (and POS) state as input. In such a case, `collectPayment` would belong to a model that cannot exist without a created order. 

I think that's one of the ideas of [RIBs architecture
](https://github.com/uber/RIBs/wiki#what-are-ribs-for)

> Minimize Global States and Decisions: Global state changes cause unpredictable behavior and can make it impossible for engineers to know the full impact of their changes. RIBs encourage encapsulating states within a deep hierarchy of well-isolated individual RIBs, thus avoiding global state issues.

I imagine such an approach would have its own set of challenges so I'm just mentioning it here for the sake of sharing thoughts.

Reply:
Thanks for these thoughts... I don't quite understand the reorganisation you mention here but that's OK, let's chat about it at some point soon. If you think it's important that we consider it now just say and I'll definitely prioritise that!

> Minimize Global States and Decisions: Global state changes cause unpredictable behavior and can make it impossible for engineers to know the full impact of their changes. RIBs encourage encapsulating states within a deep hierarchy of well-isolated individual RIBs, thus avoiding global state issues.

I suspect that POS (at least with the feature set we have now) is difficult to break down into multiple RIBs, and this would be one in a larger app in future.

That said, this is still something to aim for within POS – single source-of-truth is important, and we need to be really mindful to eliminate any overlaps in our truth storage. 

Where we have state represented by enums with associated values, it's proving to be a good way to prevent overlapping state storage which has the potential to get out of sync.

The most obvious thing to avoid are flags which store state but aren't really the source of truth for that state. I'm not going to be able to solve all that in this work, but where it makes sense I'll try to improve on it.